File: dsa/educosys/subSets.cpp
#include <iostream>
#include <vector>
using namespace std;

void helper(vector<int>& nums, int ind, vector<int>& curr, vector<vector<int>>& res)
{
    if (ind >= nums.size())
    {
        res.push_back(curr);
        return;
    }

    // include wala case
    curr.push_back(nums[ind]);
    helper(nums, ind + 1, curr, res);
    curr.pop_back();

    // exclude wala case
    helper(nums, ind + 1, curr, res);
}

vector<vector<int>> subsets(vector<int>& nums) {
    vector<int> curr;
    vector<vector<int>> res;
    helper(nums, 0, curr, res);
    return res;
}

int main(void) {
    vector<int> nums = {1, 2, 3};
    vector<vector<int>> res = subsets(nums);
    return 0;
}

File: dsa/recursion/Subsets.java
package recursion;

import java.util.ArrayList;
import java.util.List;

public class Subsets {
	
	public static void helper(int ind, int[] arr, List<Integer> curr, List<List<Integer>> res)
	{
		if (ind == arr.length)
		{
			res.add(new ArrayList<>(curr));
			return;
		}
		
		// include wala case
		curr.add(arr[ind]);
		helper(ind + 1, arr, curr, res);
		curr.remove(curr.size() - 1);
		
		// exclude wala case
		helper(ind + 1, arr, curr, res);
	}

	public static void main(String[] args) {
		int[] arr = {1, 2, 3};
		List<List<Integer>> res = new ArrayList<>();
		List<Integer> curr = new ArrayList<>();
		helper(0, arr, curr, res);
	}

}
